python/ray/dag/compiled_dag_node.py

```diff
@@ -858,6 +864,76 @@ def _get_node_id(self):
         self._dag_submitter.start()
         self._dag_output_fetcher.start()
 
+    def _verify_graph(self) -> bool:
+        """
+        Create a graph following these rules:
+
+        1. Add an edge from task_{bind_index} to task_{bind_index+1}
+           on the same actor.
+        2. Add an edge from the writer to the reader if the channel
+           isn't an NCCL channel.
+        3. Add an edge from the writer and reader of an NCCL channel
+           to the node that has the next bind index on the same actor
+           as the writer.
+
+        Use topological sort to verify whether the graph is a DAG.
+        If not, the DAG will result in a deadlock.
+        """
+
+        class GraphNode:
+            def __init__(self):
+                self.in_edges = set()
+                self.out_edges = set()
+
+            @property
+            def in_degree(self) -> int:
+                return len(self.in_edges)
+
+        from ray.dag import ClassMethodNode
+
+        def _get_next_task_idx(task: "CompiledTask") -> Optional[int]:
+            if not isinstance(task.dag_node, ClassMethodNode):
+                return None
+            actor_handle = task.dag_node._get_actor_handle()
+            bind_index = task.dag_node._get_bind_index()
+            for same_node_task in self.actor_to_tasks[actor_handle]:
+                if same_node_task.dag_node._get_bind_index() == bind_index + 1:
+                    return same_node_task.idx
+            return None
+
+        def _add_edge(
+            graph: Dict[int, GraphNode], from_idx: int, to_idx: Optional[int]
+        ):
+            if to_idx is None:
+                return
+            graph[from_idx].out_edges.add(to_idx)
+            graph[to_idx].in_edges.add(from_idx)
+
+        graph = defaultdict(GraphNode)
+        for idx, task in self.idx_to_task.items():
+            next_task_idx = _get_next_task_idx(task)
+            _add_edge(graph, idx, next_task_idx)
+            for downstream_idx in task.downstream_node_idxs:
+                if task.dag_node.type_hint.requires_nccl():
+                    _add_edge(graph, downstream_idx, next_task_idx)
+                else:
+                    _add_edge(graph, idx, downstream_idx)
+        total_edges = sum(node.in_degree for node in graph.values())
+
+        zero_in_degree_nodes = deque()
+        for idx, node in graph.items():
+            if node.in_degree == 0:
+                zero_in_degree_nodes.append(idx)
+        removed_edges = 0
+        while zero_in_degree_nodes:
```

Comment:
I found this loop a bit confusing. Can you explain the loop a bit more and why the final return value corresponds to deadlock?

(I don't know if it will be simpler but I was sort of expecting to just see something like a `visited` set to check for cycles)

Reply:
Updated to check nodes instead of edges: https://github.com/ray-project/ray/pull/45960/commits/bdfde6e0cddffec624b830299565fb6b49ec3d7c. Hope it is easier to understand now.

> Can you explain the loop a bit more and why the final return value corresponds to deadlock?

This loop is used to find the topological order. We add a node to the topological order only if its `in_degree` is 0 (meaning all prerequisites are fulfilled). Hence, all nodes are included in the topological order only if all edges in the graph should be removed.

Reply:
Thanks, it makes sense now!